iceoryx_posh/include/iceoryx_posh/internal/popo/client_impl.hpp

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_CLIENT_IMPL_HPP
+#define IOX_POSH_POPO_CLIENT_IMPL_HPP
+
+#include "iceoryx_posh/capro/service_description.hpp"
+#include "iceoryx_posh/internal/popo/base_client.hpp"
+#include "iceoryx_posh/internal/popo/request_deleter.hpp"
+#include "iceoryx_posh/internal/popo/response_deleter.hpp"
+#include "iceoryx_posh/internal/popo/rpc_interface.hpp"
+#include "iceoryx_posh/popo/client_options.hpp"
+#include "iceoryx_posh/popo/request.hpp"
+#include "iceoryx_posh/popo/response.hpp"
+#include "iceoryx_posh/popo/trigger_handle.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+namespace iox
+{
+namespace popo
+{
+template <typename Req, typename Res, typename BaseClientT = BaseClient<>>
+class ClientImpl : public BaseClientT, public RpcInterface<Request<Req>>
+{
+  public:
+    explicit ClientImpl(const capro::ServiceDescription& service, const ClientOptions& clientOptions = {}) noexcept;
+
+    template <typename... Args>
+    cxx::expected<Request<Req>, AllocationError> loan(Args&&... args) noexcept;
+
+    void send(Request<Req>&& request) noexcept override;
```

Comment:
@elBoberido I think we should check whether there is a server. If not we should return something like NOT_CONNECTED_TO_SERVER. Sure there are a lot of different states with clients and servers that can come and go, but if there is no connection to a server (which we can check with something like `port.getConnectionState()`) we should tell the user that it does not make sense to wait for responses. 
If we do not block the server queue and drop a request because of overflow, a SEVER_REQUEST_QUEUE_OVERFLOW would also be nice (not sure how easy this is but deep in the stack we have this information)

Reply:
@budrus the elf had the same suggestion. Additionally he suggested to even return the chunk in case of an overflow. This was not easily doable since it requires quite big changes in the lower layers. We wanted to change this once the functionality is merged to master but it seems that we are running out of time

Reply:
@elBoberido Returning the chunk is maybe too much. Detecting and returning an error enum would be fine for me

Reply:
@budrus okay. Since this change would affect multiple open PRs, would you be fine to do this in a follow up and change it in the port, untyped C++ API, typed C++ API and C binding at once?

Reply:
@elBoberido makes sense. So I'll resolve this thread for now